### ISSUE TITLE
Crash in context menu on image layer with pyramids

### ISSUE THREAD

**Issue by aeisenbarth** [napari/napari#3523]
## 🐛 Bug

When opening the context menu on image layers with pyramids, Napari crashes.

## To Reproduce

1. Create an image layer with pyramids:
```python
import napari
import numpy as np

width = 16
n_pyramids = 4
a0 = np.random.randint(10, size=width**2).reshape((width, width))
pyramids = [a0[::2**i, ::2**i] for i in range(1, n_pyramids)]

napari.view_image(pyramids, multiscale=True)
```
2. In the layers panel, right-click the image layer.
3. It will crash because the lambda `"ndim"` ([napari/components/layerlist.py:399](https://github.com/napari/napari/blob/90f8572e8d807677e142cbe55f0dce5547ef5920/napari/components/layerlist.py#L399)) assumes `data` is an array, not a list of arrays. The attribute reader for `ndim` defaults to `None` which can't be compared with integers.
```
WARNING: Traceback (most recent call last):
  File "…/lib/python3.8/site-packages/napari/_qt/containers/_layer_delegate.py", line 152, in editorEvent
    self.show_context_menu(
  File "…/lib/python3.8/site-packages/napari/_qt/containers/_layer_delegate.py", line 184, in show_context_menu
    self._context_menu.update_from_context(layer_list._selection_context())
  File "…/lib/python3.8/site-packages/napari/_qt/widgets/qt_action_context_menu.py", line 106, in update_from_context
    enabled = eval(d['enable_when'], {}, ctx)
  File "<string>", line 1, in <module>
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

## Expected behavior

It should not crash.

## Environment

```
napari: 0.4.11
Platform: Linux-5.4.0-87-generic-x86_64-with-glibc2.17
System: Ubuntu 18.04.6 LTS
Python: 3.8.12 (default, Oct 12 2021, 13:49:34) [GCC 7.5.0]
Qt: 5.15.2
PyQt5: 5.15.5
NumPy: 1.21.3
SciPy: 1.7.1
Dask: 2021.10.0
VisPy: 0.9.2

OpenGL:
- GL version: 4.6.0 NVIDIA 390.144
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1200, scale 1.0
- screen 2: resolution 1920x1200, scale 1.0

Plugins:
- console: 0.0.4
- scikit-image: 0.4.11
- svg: 0.1.5
```
It did not happen in napari==0.4.10, it does still happen in 0.4.12.

**Comment by tlambert03** [comment]
@jni we need something like #2683

**Comment by DragaDoncila** [comment]
I am in full support of something like #2683. This has come up a lot for me when writing plugins as well. Anything that takes a layer has to do a little check in case that layer is a list or risk erroring out, and while that check is easy enough to do, it's pretty forgettable...

I also see the point that it's up to the plugin/specific code piece to decide exactly how to handle pyramids, but we've run into this issue with `layer.data` so many times, it feels like we need to come up with a sensible default and maybe some API through which that default can be changed by the user/plugin developer if something else is needed.

**Comment by aeisenbarth** [comment]
This issue still persists and wasn't fully solved by PR #3560.

When context-clicking a layer, context menu items evaluate a condition which may check for example whether the layer's `ndim` is greater 2 (for "Make Projection"). The context available to that condition contains `"active_layer_ndim" = None`, because the layer data has `shape` but not `ndim`.

**Comment by jni** [comment]
Thanks for re-upping @aeisenbarth! I'm hoping to finish #3715 in the coming days, which hopefully will put this to rest once and for all. :crossed_fingers:

**Comment by psobolewskiPhD** [comment]
The layer contextual menus are now using app_model and I cannot reproduce this on released 0.4.18. Closing, but if someone can reproduce we can revisit!
